### PR TITLE
[ BDHK-362 ] CLI Generate x-chain signature

### PR DESCRIPTION
**Pull request by blewater**
- Synchronized implementation of single-signature generation to generate identical values with Solidity.
- Upgraded `sign` functions to produce cross-chain signatures.
- Refactored sign and verification functions to support uni- and multi-userOps signature.

## Summary by CodeRabbit

- **New Features**
	- Enhanced signing and verification process for user operations, allowing for batch processing.
	- Introduced cross-chain signing functionality for user operations.

- **Bug Fixes**
	- Improved error handling in signing tests, replacing panic calls with assertions.

- **Tests**
	- Expanded test coverage for user operation signing, including conventional and Intent user operations.
	- Added new tests for cross-chain ECDSA signature generation and verification.

### CODE REVIEW

File: pkg/userop/xchain_test.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+	publicKey := privateKey.Public().(*ecdsa.PublicKey)
+	signer := &signer.EOA{
+		PrivateKey: privateKey,
+		PublicKey:  publicKey,
+	}
+
+	entryPointAddr := common.HexToAddress("0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789")
+	sourceChainID := big.NewInt(137) // Polygon
+	destChainID := big.NewInt(56)    // BSC
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/xchain_test.go:31]: This is a test private key and matches a Solidity test key. It is mandatory to test identical signature generation.

File: pkg/userop/xchain_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+			PaymasterAndData:     []byte{},
+			Signature:            []byte{},
+		}
+	}
+
+	srcIntent := `{"chainId":137, "sender":"0x18Dd70639de2ca9146C32f9c84B90A68bBDaAA96","kind":"swap","hash":"","sellToken":"0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE","buyToken":"0xc2132D05D31c914a87C6611C10748AEb04B58e8F","sellAmount":10,"buyAmount":5,"partiallyFillable":false,"status":"Received","createdAt":0,"expirationAt":0}`
+	destIntent := srcIntent // Same intent for both chains in this example
+
+	sourceUserOp := createUserOp(srcIntent)
+	destUserOp := createUserOp(destIntent)
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/xchain_test.go:54]: The actual content is irrelevant and uses virtual test properties.

File: pkg/userop/xchain_test.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+	// Verify the signature
+	isValid := userop.VerifyXSignature(chainIDs, signer.PublicKey, entryPointAddr, signedUserOps)
+	require.True(t, isValid, "Signature is invalid for cross-chain UserOperations")
+
+	// Check if the signature matches the one from the Solidity test
+	expectedSignature := "babbe1fb0f5154d9fb263e64d1b0d9a74b184aaaaa8655a3a4fc8344bc4a8580691393eecc7cd518f0f95119c235b11c566941517aa8cc74ddf0add0af1131ae1b"
+	actualSignature := fmt.Sprintf("%x", signedUserOps[0].Signature)
+	require.Equal(t, expectedSignature, actualSignature, "Generated signature does not match expected signature")
+}
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/xchain_test.go:71]: The test title explains the significance of the comparison (Solidity test alignment).

File: pkg/userop/userop.go [hunk 4]
```diff
@@ -13,64 +13,149 @@ import (
...
+	if count > 1 {
+		// Sort the hashes
+		for i := 0; i < count; i++ {
+			for j := i + 1; j < count; j++ {
+				if hashBigs[i].Cmp(hashBigs[j]) > 0 {
+					hashBigs[i], hashBigs[j] = hashBigs[j], hashBigs[i]
+					hashes[i], hashes[j] = hashes[j], hashes[i]
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/userop.go:83]: The key phrase is "a large number of UserOperation instances." They will always remain single-digit and likely less than 4 or 5, in which case bubble comparative sorting performance is practically identical. Also note that I perform the sorting operations once for both slices `hashBigs, and hashes' vs `sort.Slice`.

File: pkg/userop/userop.go [hunk 5]
```diff
@@ -13,64 +13,149 @@ import (
...
-	if userOp.Signature[64] != 27 && userOp.Signature[64] != 28 {
+	if signature[64] != 27 && signature[64] != 28 {
 		panic(errors.New("invalid Ethereum signature (V is not 27 or 28)"))
 	}
 
-	signature := bytes.Clone(userOp.Signature) // Not in RSV format
+	messageHash, err := computeMessageHash(chainIDs, entryPointAddr, userOps)
+	if err != nil {
+		fmt.Printf("Failed to compute message hash: %v\n", err)
+		return false
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/userop.go:145]: This is a CLI tool and provides additional context.

File: pkg/userop/userop.go [hunk 6]
```diff
@@ -13,64 +13,149 @@ import (
...
+	if len(userOps) == 0 {
+		return false
+	}
+
+	signature := userOps[0].Signature
+	if len(signature) != 65 {
 		panic(errors.New("signature must be 65 bytes long"))
 	}
-	if userOp.Signature[64] != 27 && userOp.Signature[64] != 28 {
+	if signature[64] != 27 && signature[64] != 28 {
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/userop.go:138]: It is a CLI tool and this is a critical error that panic halts execution and provides a stack trace.

File: pkg/userop/userop.go [hunk 7]
```diff
@@ -13,64 +13,149 @@ import (
...
 
-	prefixedHash := crypto.Keccak256Hash(
-		[]byte(fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(userOpHash), userOpHash)),
-	)
+	prefixedHash := getEtherMsgHash(messageHash)
 
-	recoveredPubKey, err := crypto.SigToPub(prefixedHash.Bytes(), signature)
+	recoveredPubKey, err := crypto.SigToPub(prefixedHash.Bytes(), sigCopy)
 	if err != nil {
 		fmt.Printf("Failed to recover public key: %v\n", err)
```
[the remark this answers is not in the document]
Comment (blewater) [reply pkg/userop/userop.go:160]: It is a CLI tool, and Printf provides additional context.

**Review by singhhp1069 — APPROVED** [review]
LGTM!  previously might have skipped instead using `recoveredAddress == expectedAddress` in `verifySignature` , `recoveredAddress.Equal(expectedAddress)` is better

**Comment by singhhp1069** [comment]
> expectedAddress

yes, correct my bad, i meant `bytes.Equal()`

**Pull request merged**